enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,254 @@
+
+### Standard SLO Behavior
+
+The CHNSO is expected to follow the standard behavior of SLO, additionally, since the CHNSO is applicable only to clusters running on on-prem ( 'baremetal', 'openstack', 'vsphere' or 'ovirt') infrastructure,
+It should support also the expected behaviors for "not in use" SLOs similar to the [CBO operator](https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/an-slo-for-baremetal.md#standard-slo-behaviors).
+
+### cluster-hosted-net-services-operator details
+
+The CHNSO should:
+
+- Be a new `openshift/cluster-hosted-net-services-operator` project.
+- Publish an image called `cluster-hosted-net-services-operator`.
+- Add a new `ClusterOperator` with an additional
+  `Disabled` status for non on-prem platforms.
+- Implement a controller reconciling a singleton instance representing the desired cluster hosted LB/DNS/VIP configuration.
+- Do nothing except set `Disabled=true`, `Available=true`, and
+  `Progressing=False` when the `Infrastructure` resource is a platform
+  type other than openstack, baremetal, ovirt or vsphere.
+
+
+#### CRD definition
+
+Below is a possible option for a CRD instance of this operator:
+
+
+```yaml
+apiVersion: clusterhostednetservices.operator.io/v1alpha1
+kind: ClusterHostedNetServicesConfig
+metadata:
+  name: cluster
+spec:
+  dns: 
+    nodesResolution: Enable
+    appsResolution: Enable    
+    apiintIpAddress: 192.168.111.5 
+  ingressHa: Enable
+status:
+  apiServerInternalIpOwner: master-0
+  ingressIpOwner: worker-1
+```
+
+With this sample manifest the CHNSO should:
+- Create DNS hosted service that will:
+    -  resolve node names and .apps wildcard record
+    -  resolve api-int to 192.168.111.5.
+- Run the self-hosted Loadbalancer for api only if apiintIpAddress is equal to API-VIP value provides in install-config file.
+- Provide high availability for default ingress.
+
+
+From the status section we can see that master-0 node currently owns api-int VIP and worker-1 node owns ingress VIP. Additional spec and status fields may be introduced based on requirements.
+
+#### Managing  the cluster hosted networking services
+
+All the LB, DNS, and VIP components will be enabled by default in case of fresh deployment as today, while post-deployment the admin can control the services by editing the singleton CR.
+
+##### API load balancer migration
+
+-- self-hosted to an external Load balancer
+
+In order to migrate the self-hosted LB to an external Load balancer the admin should:
+ - Provide new IP address (!= API-VIP from install-config) pointing to external Load balancer front end.
+ - Edit apiintIpAddress value to use this IP address.
+ 
+As a result of this change the CHNSO should:
+- First, verify that IP address is valid and kube-api is available using the new IP address.
+- Update CoreDNS to resolve api-int to the new IP address
+- Remove the HAProxy pods after graceful shutdown timeout 
+- Keepalived static pods will continue running
```

Comment:
Be explicit that there are no requirements that this process is smooth in any way. It might be that it takes minutes with possible outage of components (kubelet, kube-scheduler, ...).

Reply:
Yep, good point. 10x